torch_em/trainer/default_trainer.py

```diff
@@ -158,93 +160,173 @@ def load_generic(
-            "mixed_precision",
-            "early_stopping",
-        ]:
-            kwargs[kw_name] = load(kw_name)
-
-        return kwargs
-
     @classmethod
     def from_checkpoint(cls, checkpoint_folder, name="best", device=None):
         save_path = os.path.join(checkpoint_folder, f"{name}.pt")
         save_dict = cls._get_save_dict(save_path, device)
         deserializer = cls.Deserializer(save_dict["init"], save_path, device)
-        trainer_kwargs = cls._get_trainer_kwargs(deserializer)
-        trainer = cls(**trainer_kwargs)
+        for name, parameter in inspect.signature(cls).parameters.items():
+            deserializer.load(name, optional=parameter.default is not inspect.Parameter.empty)
+
+        trainer = cls(**deserializer.trainer_kwargs)
         trainer._initialize(0, save_dict)
         return trainer
 
-    def _build_init(self):
-
-        def _full_class_path(obj):
-            return f"{obj.__class__.__module__}.{obj.__class__.__name__}"
-
-        def _full_class_path_of_class(obj):
-            return f"{obj.__module__}.{obj.__name__}"
-
-        def _update_loader(init_data, loader, name):
-            init_data.update({
-                f"{name}_dataset": loader.dataset,
-                f"{name}_loader_kwargs": get_constructor_arguments(loader)
-            })
-            return init_data
-
-        init_data = {
-            "model_class": _full_class_path(self.model),
-            "model_kwargs": get_constructor_arguments(self.model),
-            "loss_class": _full_class_path(self.loss),
-            "loss_kwargs": get_constructor_arguments(self.loss),
-            "optimizer_class": _full_class_path(self.optimizer),
-            "optimizer_kwargs": get_constructor_arguments(self.optimizer),
-            "metric_class": _full_class_path(self.metric),
-            "metric_kwargs": get_constructor_arguments(self.metric),
-            "device": self.device.type,
-            "log_image_interval": self.log_image_interval,
-            "mixed_precision": self.mixed_precision,
-            "early_stopping": self.early_stopping,
-            "logger_class": None if self.logger_class is None else _full_class_path_of_class(self.logger_class),
-            "logger_kwargs": self.logger_kwargs,
-        }
-        init_data = _update_loader(init_data, self.train_loader, "train")
-        init_data = _update_loader(init_data, self.val_loader, "val")
-        if self.lr_scheduler is not None:
-            init_data["lr_scheduler_class"] = _full_class_path(self.lr_scheduler)
-            init_data["lr_scheduler_kwargs"] = get_constructor_arguments(self.lr_scheduler)
-        return init_data
+    class Serializer:
+        """Implements how to serialize trainer kwargs from a trainer instance
+
+        Examples:
+            To extend the serialization process you can inherite from this Serializer in a derived Trainer class.
+            Note that `DefaultTrainer.Serializer.dump_generic()` covers most cases already.
```

Comment:
There is no `dump_generic` method, but different specializations of this. I suggest to change the text here a bit, otherwise it's a bit confusing that this refers to a method that doesn't exist.

Reply:
fixed in https://github.com/constantinpape/torch-em/pull/54/commits/68887b1125c2c80a422737780057ea93e7863b9f#